client/client.go

```diff
@@ -143,11 +145,18 @@ func FromContext(ctx context.Context) Info {
 func NewMetadata(md map[string][]string) Metadata {
 	return Metadata{
 		data: md,
+		mu:   &sync.Mutex{},
 	}
 }
 
 // Get gets the value of the key from metadata, returning a copy.
 func (m Metadata) Get(key string) []string {
+	if m.data == nil {
+		return make([]string, 0)
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
```

Comment:
There are few other issues here:
1. We don't copy the initial map, so even if we protect this that does not mean we are thread-safe.
2. I would rather remove the "optimization" and remove the lock.

Reply:
@bogdandrutu I believe if we copy the map will will still need the lock.  The new copy of the original map will live in `Metadata.data` which lives in `Info` which lives in the `context.Context`. The same `context.Context` (and therefore `Metadata.data`) is being shared between multiple exporters.  When multiple exporters use the headersetters extension to access values in the context we get the error. Since multiple goroutines are trying to access the same map I believe we need the lock.

I started with a `sync.Map` but it made some unit tests fail in contrib.  If you like that approach more I can switch back to it and deal with contrib.

Reply:
> Since multiple goroutines are trying to access the same map I believe we need the lock.

Not true, this is only because we also write (see the "optimization" in the code). Without that we will not get any error.

Reply:
Oh I see what you mean. I'll make those changes.

Reply:
@bogdandrutu let me know what you think about the new implementation